### PR TITLE
Early exit if program doesn't contain a main function

### LINKED ISSUE #1452: cargo-miri panic's on no_main binaries
While there are many other questions about what miri should do for binaries with `#![no_main]`, the current behavior is to panic. In the spirit of #1001, it would be nice explicitly indicate that miri doesn't support testing freestanding binaries at the moment.

<details>
<summary>An example backtrace</summary>
<pre>
[-bash] Wed 17 Jun 16:40 [[master $] /Volumes/code/helena-project/tock/boards/hail]
$ cargo miri test
    Checking hail v0.1.0 (/Volumes/code/helena-project/tock/boards/hail)
warning: unused import: `core::panic::PanicInfo`
 --> boards/hail/src/io.rs:2:5
  |
2 | use core::panic::PanicInfo;
  |     ^^^^^^^^^^^^^^^^^^^^^^
  |
  = note: `#[warn(unused_imports)]` on by default

warning: unused import: `cortexm4`
 --> boards/hail/src/io.rs:4:5
  |
4 | use cortexm4;
  |     ^^^^^^^^

warning: unused import: `kernel::debug`
 --> boards/hail/src/io.rs:5:5
  |
5 | use kernel::debug;
  |     ^^^^^^^^^^^^^

warning: unused import: `kernel::hil::led`
 --> boards/hail/src/io.rs:7:5
  |
7 | use kernel::hil::led;
  |     ^^^^^^^^^^^^^^^^

warning: unused import: `crate::CHIP`
  --> boards/hail/src/io.rs:10:5
   |
10 | use crate::CHIP;
   |     ^^^^^^^^^^^

warning: unused import: `crate::PROCESSES`
  --> boards/hail/src/io.rs:11:5
   |
11 | use crate::PROCESSES;
   |     ^^^^^^^^^^^^^^^^

warning: static is never used: `WRITER`
  --> boards/hail/src/io.rs:17:1
   |
17 | static mut WRITER: Writer = Writer { initialized: false };
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = note: `#[warn(dead_code)]` on by default

thread 'rustc' panicked at 'no main function found!', src/tools/miri/src/bin/miri.rs:35:37
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace

error: internal compiler error: unexpected panic

note: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/blob/master/CONTRIBUTING.md#bug-reports

note: rustc 1.45.0-nightly (fe10f1a49 2020-06-02) running on x86_64-apple-darwin

note: compiler flags: -C opt-level=z -C embed-bitcode=no -C debuginfo=2 -C debug-assertions=on -C incremental

note: some of the compiler flags provided by cargo are hidden

warning: 7 warnings emitted

error: could not compile `hail`.

To learn more, run the command again with --verbose.
</pre>
</details>

I've attached a backtrace, but any `no_main` binary will crash in a similar manner right now.

CC https://github.com/tock/tock/issues/1713

### PR DESCRIPTION
**Pull request by Stupremee**
Resolves #1452

### CODE REVIEW

**Comment by oli-obk** [comment]
This looks good to me. Can you add a test to `tests/ui/compile-fail` to check this behaviour?

File: tests/run-pass/no_main.stderr [hunk 1]
```diff
@@ -0,0 +1,2 @@
+error: miri can only run programs that have a main function
```
Comment (RalfJung) [inline tests/run-pass/no_main.stderr:1]: Hm, seems a bit odd to have this as a run-pass test.^^ I think that just happens to work as run-pass currently ignores the exit code, but IMO we should not rely on that.

Comment (Stupremee) [reply tests/run-pass/no_main.stderr:1]: Yes that makes sense, but if I move both files to the `compile-fail` folder, the test doesn't pass

Comment (oli-obk) [reply tests/run-pass/no_main.stderr:1]: What's the test failure that you're getting?

File: tests/run-pass/no_main.rs [hunk 2]
```diff
@@ -0,0 +1 @@
+#![no_main]
```
Comment (oli-obk) [inline tests/run-pass/no_main.rs:1]: you may be able to make this test pass in `compile-fail` by adding a `// error-pattern miri can only run programs that have a main function` comment to this file

Comment (Stupremee) [reply tests/run-pass/no_main.rs:1]: Ah yes that works :+1:

**Comment by oli-obk** [comment]
@bors r+

Thanks!

**Comment by RalfJung** [comment]
Thanks a lot for the patch @Stupremee :)

**Pull request merged**